File: tools/aengine-editor/core/src/com/aengineeditor/game/SceneObject.java
package com.aengineeditor.game;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Input;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.VertexAttributes;
import com.badlogic.gdx.graphics.g3d.*;
import com.badlogic.gdx.graphics.g3d.attributes.ColorAttribute;
import com.badlogic.gdx.graphics.g3d.attributes.TextureAttribute;
import com.badlogic.gdx.graphics.g3d.utils.ModelBuilder;
import com.badlogic.gdx.math.Quaternion;
import com.badlogic.gdx.math.Vector3;
import com.badlogic.gdx.math.collision.BoundingBox;
import com.badlogic.gdx.math.collision.Ray;

public class SceneObject {
    public Vector3 position;
    public Vector3 scale;
    public Vector3 rotation;
    public int id;

    public boolean selected;

    Model model;
    ModelInstance instance;
    BoundingBox boundingBox;

    Model outline;
    ModelInstance outlineInstance;

    public SceneObject(int id, Color color) {
        position = new Vector3();
        scale = new Vector3(2, 2, 2);
        rotation = new Vector3();
        this.id = id;
        selected = false;

        ModelBuilder modelBuilder = new ModelBuilder();
        model = modelBuilder.createBox(scale.x, scale.y, scale.z, new Material(ColorAttribute.createDiffuse(color)), VertexAttributes.Usage.Position | VertexAttributes.Usage.Normal);
        instance = new ModelInstance(model);

        outline = modelBuilder.createBox(scale.x, scale.y, scale.z, new Material(TextureAttribute.createDiffuse(new Texture(Gdx.files.internal("logo.png")))), VertexAttributes.Usage.Position | VertexAttributes.Usage.Normal | VertexAttributes.Usage.TextureCoordinates);
        outlineInstance = new ModelInstance(outline);

        boundingBox = new BoundingBox();
    }

    public void update() {
        if (selected) {
            if (Config.currentMode == Mode.MOVE) {
                if (Gdx.input.isKeyPressed(Input.Keys.UP)) {
                    position.x += Config.moveSpeed * Gdx.graphics.getDeltaTime();
                }
                if (Gdx.input.isKeyPressed(Input.Keys.DOWN)) {
                    position.x -= Config.moveSpeed * Gdx.graphics.getDeltaTime();
                }
                if (Gdx.input.isKeyPressed(Input.Keys.LEFT)) {
                    position.z -= Config.moveSpeed * Gdx.graphics.getDeltaTime();
                }
                if (Gdx.input.isKeyPressed(Input.Keys.RIGHT)) {
                    position.z += Config.moveSpeed * Gdx.graphics.getDeltaTime();
                }
                if (Gdx.input.isKeyPressed(Input.Keys.PAGE_UP)) {
                    position.y += Config.moveSpeed * Gdx.graphics.getDeltaTime();
                }
                if (Gdx.input.isKeyPressed(Input.Keys.PAGE_DOWN)) {
                    position.y -= Config.moveSpeed * Gdx.graphics.getDeltaTime();
                }
            }
            if (Config.currentMode == Mode.SCALE) {
                if (Gdx.input.isKeyPressed(Input.Keys.UP)) {
                    scale.x += Config.moveSpeed * Gdx.graphics.getDeltaTime();
                }
                if (Gdx.input.isKeyPressed(Input.Keys.DOWN)) {
                    scale.x -= Config.moveSpeed * Gdx.graphics.getDeltaTime();
                }
                if (Gdx.input.isKeyPressed(Input.Keys.LEFT)) {
                    scale.z -= Config.moveSpeed * Gdx.graphics.getDeltaTime();
                }
                if (Gdx.input.isKeyPressed(Input.Keys.RIGHT)) {
                    scale.z += Config.moveSpeed * Gdx.graphics.getDeltaTime();
                }
                if (Gdx.input.isKeyPressed(Input.Keys.PAGE_UP)) {
                    scale.y += Config.moveSpeed * Gdx.graphics.getDeltaTime();
                }
                if (Gdx.input.isKeyPressed(Input.Keys.PAGE_DOWN)) {
                    scale.y -= Config.moveSpeed * Gdx.graphics.getDeltaTime();
                }
            }
        }
        boundingBox.set(position, scale);
        instance.transform.set(position, new Quaternion(), scale);
        outlineInstance.transform.set(position, new Quaternion(), new Vector3(scale.x + 0.1f, scale.y + 0.1f, scale.z + 0.1f));

        System.out.println(boundingBox.toString());
    }

    public void render(ModelBatch batch, Environment environment) {
        batch.render(instance, environment);
        if (selected) batch.render(outlineInstance, environment);
    }

    public void dispose() {
        model.dispose();
    }

}

File: tools/aengine-editor-c/src/scene_object.cpp
#include "headers/scene_object.h"
#include "../include/raylib.h"
#include "iostream"

#define R_MULTIPLIER 40
#define G_MULTIPLIER 50
#define B_MULTIPLIER 60

scene_object::scene_object(int id) {
    position = {0, 0, 0};
    scale = {2, 2, 2};
    rotation = {0, 0, 0};
    this->id = id;
    selected = false;
    color = Color {(unsigned char)(R_MULTIPLIER * id), (unsigned char)(G_MULTIPLIER * id), (unsigned char)(B_MULTIPLIER * id), 255};
}

scene_object::scene_object(int id, Vector3 position) {
    this->position = position;
    scale = {2, 2, 2};
    rotation = {0, 0, 0};
    this->id = id;
    selected = false;
    color = Color {(unsigned char)(R_MULTIPLIER * id), (unsigned char)(G_MULTIPLIER * id), (unsigned char)(B_MULTIPLIER * id), 255};
}

void scene_object::update() {
    if (selected) {
        if (config::currentMode == MOVE)
        {
            if (IsKeyDown(KEY_W))
            {
                position.x += config::moveSpeed * GetFrameTime();
            }
            if (IsKeyDown(KEY_S))
            {
                position.x -= config::moveSpeed * GetFrameTime();
            }
            if (IsKeyDown(KEY_A))
            {
                position.z -= config::moveSpeed * GetFrameTime();
            }
            if (IsKeyDown(KEY_D))
            {
                position.z += config::moveSpeed * GetFrameTime();
            }
            if (IsKeyDown(KEY_PAGE_UP))
            {
                position.y += config::moveSpeed * GetFrameTime();
            }
            if (IsKeyDown(KEY_PAGE_DOWN))
            {
                position.y -= config::moveSpeed * GetFrameTime();
            }
        }
        if (config::currentMode == SCALE)
        {
            if (IsKeyDown(KEY_W))
            {
                scale.x += config::moveSpeed * GetFrameTime();
            }
            if (IsKeyDown(KEY_S))
            {
                scale.x -= config::moveSpeed * GetFrameTime();
            }
            if (IsKeyDown(KEY_A))
            {
                scale.z -= config::moveSpeed * GetFrameTime();
            }
            if (IsKeyDown(KEY_D))
            {
                scale.z += config::moveSpeed * GetFrameTime();
            }
            if (IsKeyDown(KEY_PAGE_UP))
            {
                scale.y += config::moveSpeed * GetFrameTime();
            }
            if (IsKeyDown(KEY_PAGE_DOWN))
            {
                scale.y -= config::moveSpeed * GetFrameTime();
            }
        }
    }
}

void scene_object::render() {
    DrawCubeV(position, scale, color);

    if (selected) {
        DrawCubeWiresV(position, scale, YELLOW);
    } else {
        DrawCubeWiresV(position, scale, WHITE);
    }
}